### ISSUE TITLE
Assertion Error (data != elementData_.end()) When Running SPE1CASE2 in Parallel

### ISSUE THREAD

**Issue by bska** [OPM/opm-simulators#2478]
Hi,

While trying to debug another issue I happened upon the failure/backtrace below.  This was running
```sh
$ mpirun -np 2 --map-by core --bind-to core --report-bindings \
    ${flow} --ecl-deck-file-name=${cse} --output-dir=./master.dbg
```
with `${flow}` being the full path to an unchanged (master, debug, mpi, Dune 2.6.0) build on my Ubuntu 18.04 laptop (GCC 7.5.0, OpenMPI 2.1.1-8) and `${cse}` being the path to [SPE1CASE2.DATA](https://github.com/OPM/opm-tests/blob/b08e92a94712b8378cfd8ebfbeddc070a1f3badf/spe1/SPE1CASE2.DATA).

I am fully prepared to accept that there's something wrong with my setup, but I thought I'd report it in case anyone else is seeing anything similar.  Just for the record, this runs through without issue if I switch to using `-np 1` in the `mpirun` invocation.  Sadly, my experience debugging MPI-related issues is very shallow, but I'd be happy to try all suggestions.

--- 

```
[0] Zoltan ERROR in Zoltan_Comm_Do_Post (line 180 of /build/trilinos-2G9GDl/trilinos-12.12.1/packages/zoltan/src/Utilities/Communication/comm_do.c):  nrecvs not zero, but recv_data = NULL
[0] Zoltan ERROR in Zoltan_PHG_Build_Hypergraph (line 131 of /build/trilinos-2G9GDl/trilinos-12.12.1/packages/zoltan/src/phg/phg_build.c):  Error getting graph from application
[0] Zoltan ERROR in Zoltan_PHG (line 291 of /build/trilinos-2G9GDl/trilinos-12.12.1/packages/zoltan/src/phg/phg.c):  Error building hypergraph.
[0] Zoltan ERROR in Zoltan_PHG (line 506 of /build/trilinos-2G9GDl/trilinos-12.12.1/packages/zoltan/src/phg/phg.c):  Error partitioning hypergraph.
[0] Zoltan ERROR in Zoltan_LB (line 480 of /build/trilinos-2G9GDl/trilinos-12.12.1/packages/zoltan/src/lb/lb_balance.c):  Partitioning routine returned code -1.
flow: /home/bska/work/opm/src/opm.master/opm-simulators/opm/simulators/utils/PropsCentroidsDataHandle.hpp:153: Opm::PropsCentroidsDataHandle<Grid>::~PropsCentroidsDataHandle() [with Grid = Dune::CpGrid]: Assertion `data != elementData_.end()' failed.
[bard-HP-ENVY-TS-15:19482] *** Process received signal ***
[bard-HP-ENVY-TS-15:19482] Signal: Aborted (6)
[bard-HP-ENVY-TS-15:19482] Signal code:  (-6)
[bard-HP-ENVY-TS-15:19482] [ 0] /lib/x86_64-linux-gnu/libpthread.so.0(+0x12890)[0x7fd6f353d890]
[bard-HP-ENVY-TS-15:19482] [ 1] /lib/x86_64-linux-gnu/libc.so.6(gsignal+0xc7)[0x7fd6f3178e97]
[bard-HP-ENVY-TS-15:19482] [ 2] /lib/x86_64-linux-gnu/libc.so.6(abort+0x141)[0x7fd6f317a801]
[bard-HP-ENVY-TS-15:19482] [ 3] /lib/x86_64-linux-gnu/libc.so.6(+0x3039a)[0x7fd6f316a39a]
[bard-HP-ENVY-TS-15:19482] [ 4] /lib/x86_64-linux-gnu/libc.so.6(+0x30412)[0x7fd6f316a412]
[bard-HP-ENVY-TS-15:19482] [ 5] /home/bska/work/opm/build/opm.master/normal/dbg/mpi/dune-2.6.0/inst/bin/flow(Opm::PropsCentroidsDataHandle<Dune::CpGrid>::~PropsCentroidsDataHandle()+0x3e9)[0x561fe45b7891]
[bard-HP-ENVY-TS-15:19482] [ 6] /home/bska/work/opm/build/opm.master/normal/dbg/mpi/dune-2.6.0/inst/bin/flow(Opm::EclCpGridVanguard<Opm::Properties::TTag::EclFlowProblem>::loadBalance()+0x89a)[0x561fe45880d8]
[bard-HP-ENVY-TS-15:19482] [ 7] /home/bska/work/opm/build/opm.master/normal/dbg/mpi/dune-2.6.0/inst/bin/flow(Opm::Simulator<Opm::Properties::TTag::EclFlowProblem>::Simulator(bool)+0x517)[0x561fe4572673]
[bard-HP-ENVY-TS-15:19482] [ 8] /home/bska/work/opm/build/opm.master/normal/dbg/mpi/dune-2.6.0/inst/bin/flow(Opm::FlowMainEbos<Opm::Properties::TTag::EclFlowProblem>::setupEbosSimulator(bool)+0x4f)[0x561fe4559ff1]
[bard-HP-ENVY-TS-15:19482] [ 9] /home/bska/work/opm/build/opm.master/normal/dbg/mpi/dune-2.6.0/inst/bin/flow(Opm::FlowMainEbos<Opm::Properties::TTag::EclFlowProblem>::execute(int, char**, bool, bool)+0x9a)[0x561fe454c044]
[bard-HP-ENVY-TS-15:19482] [10] /home/bska/work/opm/build/opm.master/normal/dbg/mpi/dune-2.6.0/inst/bin/flow(Opm::flowEbosBlackoilMain(int, char**, bool, bool)+0x8c)[0x561fe44dec1c]
[bard-HP-ENVY-TS-15:19482] [11] /home/bska/work/opm/build/opm.master/normal/dbg/mpi/dune-2.6.0/inst/bin/flow(main+0x130b)[0x561fe438c5e3]
[bard-HP-ENVY-TS-15:19482] [12] /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xe7)[0x7fd6f315bb97]
[bard-HP-ENVY-TS-15:19482] [13] /home/bska/work/opm/build/opm.master/normal/dbg/mpi/dune-2.6.0/inst/bin/flow(_start+0x2a)[0x561fe438916a]
[bard-HP-ENVY-TS-15:19482] *** End of error message ***
--------------------------------------------------------------------------
mpirun noticed that process rank 0 with PID 0 on node bard-HP-ENVY-TS-15 exited on signal 6 (Aborted).
--------------------------------------------------------------------------
```

**Comment by blattms** [comment]
Works for me.

Did you compile Zoltan yourself, what options? Wasn't there this misconfigured MPI issue on Ubuntu? Couls that have kicked in.

Or they might do additional checks when compiled in debug mode and loadbalancing fails because of that.

**Comment by bska** [comment]
> Did you compile Zoltan yourself

I didn't.  This is the Ubuntu package 'libtrilinos-zoltan-dev', version 12.12.1-5.  I can try to build Zoltan manually.

> Or they might do additional checks when compiled in debug mode and loadbalancing fails because of that.

I get the same failure, apart from the `assert` diagnostic, if I run a release/optimised build.  

> Wasn't there this misconfigured MPI issue on Ubuntu? Couls that have kicked in.

Possibly.  I'll report back when I have a self-built version of Zoltan as the backend.

**Comment by bska** [comment]
Closing.  I found a local work-around (building a local copy of the OpenMPI runtime libraries).
